### PR TITLE
Using AWS CodePipeline as CI/CD solution for MuleSoft Apps

### PR DESCRIPTION
**Pull request by atteheino**

### CODE REVIEW

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+A typical tooling setup for a integration project consists of a Git repository somewhere, Jenkins on some host, Application Server either on-premise or hosted somewhere. Git stores the source code, Jenkins builds and deploys the code to the application server running the application. 
+
+I was given the option to try out AWS alternative to this setup. AWS has [CodeCommit](https://aws.amazon.com/codecommit/) for source control, [CodeBuild](https://aws.amazon.com/codebuild/) for building your software and [CodePipeline](https://aws.amazon.com/codepipeline/) for automating the whole CI/CD process.
+
+Typically it would be a good idea to document your infrastructure as code, for example by creating the services you need via CloudFormation templates. Here I was doing a Proof Of Concept, so I did all of the setup via the AWS Console UI.
+
+# CodeCommit to store the source code
+
+It's not a requirement to use AWS CodeCommit, to be able to utilize CodeBuild or CodePipeline, but in my case it made sense as the original source code was not in a public repository like GitHub.
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:24]: Commas in this sentence are not necessary in my opinion, at least the one after CodeCommit.

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+To be able to access your new repository:
+- Create a IAM user for CodeCommit access. Grant it a suitable policy to work with CodeCommit.
+- Create HTTPS Git Credentials for Git to access the CodeCommit repository. (You can do this from IAM user's Security Credentials tab)
+- Configure Git on your computer to access CodeCommit, by using the credentials created on the previous step.
+
+The rest is pretty much trivial by following the instructions on the guide.
+
+# CodePipeline to run the entire process
+
+After having the sources in CodeCommit, I took a look at the documentation on how to proceed. I figured out, that I would first create the CodePipeline that is triggered when ever the sources change in the Git repository. Second step would be to create the CodeBuild step and it's simple to add while building the CodePipeline. The CodePipeline I created is really simple:
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:38]: Unnecessary comma between "out, that"
when ever -> whenever

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+![CodePipeline example](/img/aws-developer-tools-mule-cicd-solution/pipeline.png)
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:40]: Image is missing but is it just not working in GitHub preview and is actually fine...?

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+Just two steps. One for reacting to the source code change, and one for building the software.
+
+There are several sources for the CodePipeline that one can choose from. AWS S3, AWS CodeCommit, GitHub, Bitbucket etc.
+You can also choose the Artifact Store and encryption key for the Artifacts stored in the store.
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:45]: This sentence is a bit weird. Could it be simplified somehow? Maybe "You can also choose Artifact Store and store the needed encryptions keys for the artifacts there."?

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+You must also create a Service Worker Role or use a existing one to run the pipeline.
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:46]: a existing -> an existing

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+There is also possibility to define a Deploy Stage, but it's mostly about deploying to AWS infra, so it was not for me.
+
+# CodeBuild to build the software
+
+Final step is to use AWS CodeBuild to run Maven to build the MuleSoft app. MuleSoft apps use Maven by default, so it's a clear choise to use.
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:52]: choise -> choice

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+In the *build*-section the commands are listed that actually perform the build. For the `mvn package deploy` step the parameters come from the variables defined in the env-section. In the builds stage, I first build and package a custom connector that is used by the MuleSoft application internally.
+
+### artifacts-section
+
+Here is where you can spcify artifacts from the build process to be stored in the S3 bucket. These can then later on be used in deployments to other Amazon Services or other services that can download from S3.
+
+### cache-section
+
+As we all know, Maven tends to download half of the internet to do a build. To speed up builds we should setup a cache. See the [Cache for Maven repository](#cache-for-maven-repository) section to see how to set it up.
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:130]: I would avoid using phrases such as "As we all know". Reader might not know it. You could convey the same meaning by saying for example "Notoriously, Maven tends to download half of the internet to do a build."

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+## Secrets
+
+Storing secrets and providing them to the pipeline and build process is something one must almost always do. Documentation guides the developer to use AWS Systems Manager Parameter Store. CodeBuild is then able to access the secrets during runtime. 
+
+After adding the secrects to the Parameter Store, you must add *ssm:GetParameters* permission to the CodeBuild Service Worker role for it to be able to access the secrets. The easiest way to do so is by adding *AmazonSSMReadOnlyAccess* policy for the CodeBuild Service Worker Role.
+
+Maven Secrets are also used to encrypt the credentials needed to access the private Maven repository of MuleSoft and the deployment keys to CloudHub. The process is explained [here](https://docs.mulesoft.com/mule-runtime/3.9/mule-maven-plugin#encrypting-credentials). The actual Maven Master Password needs to be stored somewhere securely. I used AWS Secrects Manager, but any Password manager would do. 
+
+### Cache for Maven repository
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:142]: I think this section could be part of the earlier cache section, so it wouldn't then be necessary to reference between the sections.

**Review by Juholei — COMMENTED** [review]
I added a few grammar and style related comments. The post explained the things in a clear way and it was easy to follow. I would like to see blog post still be expanded a little bit with what you learned and what you think of using these AWS services. Now the blog post ends abruptly. So I suggest you compare the experience of using this approach to the traditional approach you mention in the beginning and then give some opinions when you would use this, what was good, what was bad etc. Maybe it is as simple as saying that you should use these tools if your project will be in AWS, but it adds some opinions and insight into the blog post

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md (hunk 3 above)
Comment (atteheino) [reply _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:40]: It works in the Docker preview, so I think it's ok...

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md (hunk 4 above)
Comment (atteheino) [reply _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:45]: I'll try to say it another way.. But you actually select the Artifact store location (S3 bucket) and the encryption key that is used to encrypt the artifacts that are stored to the S3 bucket..

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md (hunk 7 above)
Comment (atteheino) [reply _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:130]: Good point!

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md (hunk 8 above)
Comment (atteheino) [reply _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:142]: I agree.

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 9]
```diff
@@ -35,25 +35,27 @@ The rest is pretty much trivial by following the instructions on the guide.
...
 
 Just two steps. One for reacting to the source code change, and one for building the software.
 
-There are several sources for the CodePipeline that one can choose from. AWS S3, AWS CodeCommit, GitHub, Bitbucket etc.
-You can also choose the Artifact Store and encryption key for the Artifacts stored in the store.
-You must also create a Service Worker Role or use a existing one to run the pipeline.
+Here are a few things you must do and can do to configure the CodePipeline:
+- You must create a Service Worker Role or use an existing one to run the pipeline.
+- You can select the source for the CodePipeline. There are several sources for the CodePipeline that one can choose from: AWS S3, AWS CodeCommit, GitHub, Bitbucket etc.
+- You can select the Artifact Store. Artifact Store is simply a AWS S3 bucket. Use the default artifact store, create a new one or use an existing bucket.
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:47]: a AWS S3 bucket -> an AWS S3 bucket
Same thing furher down in the post

**Review by Juholei — COMMENTED** [review]
I added a couple grammar fixes now. I think the new summary section was good and together with that and combining the sections I mentioned earlier the flow of the blog post is much better now. One thing I noted now that the deployment target, CloudHub is mentioned but not introduced. I didn't know what it was before and it is a large part of the whole thing, so I think it should be introduced in the beginning briefly with a sentence or two, alongside the AWS services used. You could then also add a short overview in the beginning what it is that we are going to do in this blog post. Something like "code comes from repo, CodePipeline defines what is done to it, CodeBuild builds it and then it it is deployed to CloudHub", or however it goes.

File: _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md [hunk 10]
```diff
@@ -19,7 +19,7 @@ I was given the option to try out AWS alternative to this setup. AWS has [CodeCo
 
 Typically it would be a good idea to document your infrastructure as code, for example by creating the services you need via CloudFormation templates. Here I was doing a Proof Of Concept, so I did all of the setup via the AWS Console UI.
 
-I'm going to show here how to build a CI/CD pipeline using AWS Developer tools, utilize Maven for the actual builing and deploying the MuleSoft Application to CloudHub. CloudHub is the PaaS (Platform as a Service) part of MuleSoft's Anypoint Platform. Technically it uses AWS services in the background. Find more information [here](https://www.mulesoft.com/platform/saas/cloudhub-ipaas-cloud-based-integration).
+I'm going to show here how to build a CI/CD pipeline using AWS Developer tools and how to utilize Maven for the actual builing & deploying the MuleSoft Application to CloudHub. CloudHub is the PaaS (Platform as a Service) part of MuleSoft's Anypoint Platform. Technically it uses AWS services in the background. Find more information [here](https://www.mulesoft.com/platform/saas/cloudhub-ipaas-cloud-based-integration).
```
Comment (Juholei) [inline _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md:22]: builing -> building

Comment (atteheino) [reply _posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md]: fixed

**Pull request merged**